### PR TITLE
PMP: Add a visitor to the orientation function 

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes

Add a visitor to the orientation function and use it to fix the plugin about NM-vertices
## Release Management

* Affected package(s):PMP
* Feature/Small Feature (if any):
* Link to compiled documentation (obligatory for small feature) [*wrong link name to be changed*](httpssss://wrong_URL_to_be_changed/Manual/Pkg)
* License and copyright ownership:

### CODE REVIEW

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h [hunk 1]
```diff
@@ -63,6 +71,7 @@ struct Polygon_soup_orienter
   Edge_map edges;             //< the set of edges of the input polygons
   Marked_edges marked_edges;  //< the set of singular edges or edges incident
                               //<   to non-compatible orientation polygons
+  Visitor& visitor;
```
Comment (afabri) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h:74]: @sloriot do we want a reference or a copy?

Comment (maxGimeno) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h]: This is now a pointer.

Comment (lrineau) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h]: How is that documented?

Comment (maxGimeno) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h]: It isn't yet. I'd like to have the code reviewed first, then if we agree on the implem, I'll doc it and make a SF

Comment (lrineau) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h]: But how is it documented for other visitors in PMP? Holding a ref or a pointer opens the risk of a dandling pointer/ref. If the implementation stays as it is, then in the main function the parameter for the visitor should be a non-const ref.

Comment (maxGimeno) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h]: In PMP visitors are passed through NPs and copied.

Comment (afabri) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h]: If we do it as in the BGL, visitors are copied and hence must be lightweight.

**Comment by sloriot** [comment]
Is it ready for review?

**Comment by maxGimeno** [comment]
Review of the code yes, I didn't write any documentation yet.

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h [hunk 2]
```diff
@@ -37,7 +40,15 @@ namespace Polygon_mesh_processing {
 
 namespace internal {
 
-template<class PointRange, class PolygonRange>
+struct Polygon_soup_orientation_visitor{
+  inline virtual void non_manifold_edge(const std::size_t&, const std::size_t&){}
+  inline virtual void duplicate_vertex(const std::size_t&, const std::size_t&){}
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h:45]:
```suggestion
  inline virtual void duplicated_vertex(const std::size_t&, const std::size_t&){}
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h [hunk 3]
```diff
@@ -37,7 +40,15 @@ namespace Polygon_mesh_processing {
...
+  inline virtual void non_manifold_vertex(const std::size_t&){}
+  inline virtual void update_polygon_id(const std::size_t&, const std::size_t&, const std::size_t&){}
+  inline virtual void reverse_polygon(const std::size_t&) {}
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h:48]:
```suggestion
  inline virtual void polygon_orientation_reversed(const std::size_t&) {}
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h [hunk 4]
```diff
@@ -37,7 +40,15 @@ namespace Polygon_mesh_processing {
 
 namespace internal {
 
-template<class PointRange, class PolygonRange>
+struct Polygon_soup_orientation_visitor{
+  inline virtual void non_manifold_edge(const std::size_t&, const std::size_t&){}
+  inline virtual void duplicate_vertex(const std::size_t&, const std::size_t&){}
+  inline virtual void non_manifold_vertex(const std::size_t&){}
+  inline virtual void update_polygon_id(const std::size_t&, const std::size_t&, const std::size_t&){}
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h:47]:
```suggestion
  inline virtual void point_id_in_polygon_updated(const std::size_t&, const std::size_t&, const std::size_t&){}
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h [hunk 5]
```diff
@@ -302,7 +331,7 @@ struct Polygon_soup_orienter
     } // end while loop on all non-oriented polygons remaining
   }
 
-  /// A vertex is said to be singular if its link is neither a cycle nor a chain,
+  /// A vertex is said to be .3 if its link is neither a cycle nor a chain,
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/orient_polygon_soup.h:334]: typo

File: Polyhedron/demo/Polyhedron/Plugins/PMP/Orient_soup_plugin.cpp [hunk 6]
```diff
@@ -254,18 +252,46 @@ void Polyhedron_demo_orient_soup_plugin::createPointsAndPolyline()
   if(item)
   {
     QApplication::setOverrideCursor(Qt::WaitCursor);
-    Scene_points_with_normal_item* points = new Scene_points_with_normal_item();
+    Scene_points_with_normal_item* points = nullptr;
+
+    //remove vertices already in NM edges
+
+    std::vector<std::pair<std::size_t, std::size_t> > to_remove;
```
Comment (sloriot) [inline Polyhedron/demo/Polyhedron/Plugins/PMP/Orient_soup_plugin.cpp:259]: This part seems overly complicated.

File: Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp [hunk 7]
```diff
@@ -420,9 +420,23 @@ void Scene_polygon_soup_item::inside_out()
...
+Scene_polygon_soup_item::orient(std::vector<std::pair<std::size_t, std::size_t> >& non_manifold_vertices)
 {
+  struct Visitor : public CGAL::Polygon_mesh_processing::internal::Polygon_soup_orientation_visitor
+  {
+    std::vector<std::pair<std::size_t, std::size_t> >& nm_vertices;
+
+    Visitor(std::vector<std::pair<std::size_t, std::size_t> >& nm_vertices)
+      :nm_vertices(nm_vertices){}
+
+    Visitor(const Visitor& other)
```
Comment (sloriot) [inline Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp:432]: do you really need that?

Comment (sloriot) [reply Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp:432]: at least ` = default` should be better.

File: Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp [hunk 8]
```diff
@@ -420,9 +420,23 @@ void Scene_polygon_soup_item::inside_out()
 }
 
 bool
-Scene_polygon_soup_item::orient()
+Scene_polygon_soup_item::orient(std::vector<std::pair<std::size_t, std::size_t> >& non_manifold_vertices)
 {
+  struct Visitor : public CGAL::Polygon_mesh_processing::internal::Polygon_soup_orientation_visitor
+  {
+    std::vector<std::pair<std::size_t, std::size_t> >& nm_vertices;
```
Comment (sloriot) [inline Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp:427]: does the demo require a pair?

Comment (maxGimeno) [reply Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp:427]: Yes. To separate NM vertices from NM Edges endpoints, you have to check if a nm vertex is one of the endpoints of an edge. But you cannot know which vertex you are comparing, as it has been duplicated, so you have to compare with both, hence the pair.

File: Polyhedron/demo/Polyhedron/Plugins/PMP/Orient_soup_plugin.cpp (hunk 6 above)
Comment (maxGimeno) [reply Polyhedron/demo/Polyhedron/Plugins/PMP/Orient_soup_plugin.cpp:259]: And I was so proud of it :(

File: Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp (hunk 7 above)
Comment (maxGimeno) [reply Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp:432]: You're right, the default is fine.

File: Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp (hunk 8 above)
Comment (sloriot) [reply Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp:427]: can't you simply insert both in your container?

File: Polyhedron/demo/Polyhedron/Plugins/PMP/Orient_soup_plugin.cpp (hunk 6 above)
Comment (sloriot) [reply Polyhedron/demo/Polyhedron/Plugins/PMP/Orient_soup_plugin.cpp:259]: Maybe I missed something but to me you only have to collect points that are reported as nm but that are not reported in an edge.

Comment (sloriot) [reply Polyhedron/demo/Polyhedron/Plugins/PMP/Orient_soup_plugin.cpp:259]: Oh but maybe that's the thing. nm edges are only reported using input indices?

Comment (maxGimeno) [reply Polyhedron/demo/Polyhedron/Plugins/PMP/Orient_soup_plugin.cpp:259]: Yes

File: Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp (hunk 8 above)
Comment (maxGimeno) [reply Polyhedron/demo/Polyhedron/Scene_polygon_soup_item.cpp:427]: If one of the members of the pair matches, both members need to be removed, so I need the pair

**Comment by sloriot** [comment]
I think we should have a function `non_manifold_vertex()` that reports each non-manifold vertex (edge and vertex only as already discussed), and do not only rely on `duplicated_vertex()`. I think all vertex should be replaced by point. @MaelRL, do you have an opinion?

**Comment by sloriot** [comment]
`PMPOrientationVisitor` -> `PMPSoupOrientationVisitor` ?

File: Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPOrientationVisitor.h [hunk 9]
```diff
@@ -0,0 +1,43 @@
+/// \ingroup PkgPolygonMeshProcessingConcepts
+/// \cgalConcept
+///
+/// The concept `PMPOrientationVisitor` defines the requirements for the visitor
+/// used in `CGAL::Polygon_mesh_processing::orient_polygon_soup()` to be notified
+/// of the presence of non-manifold simplices and of the modifications done to polygons
+/// during the orientation process.
```
Comment (MaelRL) [inline Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPOrientationVisitor.h:7]:
```suggestion
/// The concept `PMPOrientationVisitor` defines the requirements for the visitor
/// used in `CGAL::Polygon_mesh_processing::orient_polygon_soup()` to track
/// the detection of non-manifold simplices and the modifications performed to polygons
/// during the orientation process.
```

File: Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPOrientationVisitor.h [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  /// Non-manifoldness is resolved in the algorithm by duplicating the point.
+  /// `input_id` is the index of the input point, and `new_id` is the index of the new point.
+  /// Note that a point might be duplicated several times.
+  void duplicated_vertex(const std::size_t& input_id, const std::size_t& new_id);
+  /// is called when a the point with id `input_id` in polygon with id `pid` is replaced
+  /// by the point with id `new_id`. This functions is called after all calls to `duplicated_vertex()` are done.
+  void point_id_in_polygon_updated(const std::size_t& pid, const std::size_t& input_id, const std::size_t& new_id);
+/// @}
+
+};
```
Comment (MaelRL) [inline Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPOrientationVisitor.h:40]:
```suggestion
/// @name Functions used to report non-manifold simplices.
/// @{

/// called each time an edge appears in more than two polygons.
/// `id1` and `id2` are the ids of the endpoints of the edge in the input point range.
void non_manifold_edge(const std::size_t & id1, const std::size_t& id2);

/// called each time a non-manifold vertex is detected.
/// `id`is the id of the non-manifold vertex.
void non_manifold_vertex(const std::size_t & id);

/// @}

/// @name Functions used to report modifications done to the polygons.
/// @{

/// called when the orientation of a polygon is not compatible with its neighbors
/// and the polygon is reversed. `id` is the index of the polygon in the input polygon range.
void polygon_orientation_reversed(const std::size_t& id) ;

/// called for each non-manifold vertex (whether part of a non-manifold edge or not).
/// Non-manifoldness is resolved in the algorithm by duplicating the point.
/// `input_id` is the index of the input point, and `new_id` is the index of the new point.
/// Note that a point might be duplicated several times.
void duplicated_vertex(const std::size_t& input_id, const std::size_t& new_id);

/// called when the point with id `input_id` in polygon with id `pid` is replaced
/// by the point with id `new_id`. This function is called after all calls to `duplicated_vertex()`.
void point_id_in_polygon_updated(const std::size_t& pid, const std::size_t& input_id, const std::size_t& new_id);
  
/// @}
};
```

File: Polygon_mesh_processing/examples/Polygon_mesh_processing/orient_polygon_soup_example.cpp [hunk 11]
```diff
@@ -16,6 +16,31 @@
...
+  }
+  void point_id_in_polygon_updated(const std::size_t& p_id, const std::size_t& i1, const std::size_t& i2) final
+  {
+    std::cout<<"In the polygon "<<p_id<<"< the index "<<i1<<" has been replaced by "<<i2<<"."<<std::endl;
+  }
+  void polygon_orientation_reversed(const std::size_t& p_id) final
+  {
+    std::cout<<"The polygon "<<p_id<<" has been reversed."<<std::endl;
+  }
+  };
```
Comment (MaelRL) [inline Polygon_mesh_processing/examples/Polygon_mesh_processing/orient_polygon_soup_example.cpp:42]:
```suggestion
  void non_manifold_edge(const std::size_t& id1, const std::size_t& id2) final
  {
    std::cout << "The edge " << id1 << ", " << id2 << " is not manifold." << std::endl;
  }
  void non_manifold_vertex(const std::size_t & id) final
  {
     std::cout << "The vertex " << id << " is not manifold." << std::endl;
  }
  void duplicated_vertex(const std::size_t& v1, const std::size_t& v2) final
  {
    std::cout << "The vertex " << v1 << " has been duplicated, its new id is " << v2 << "." << std::endl;
  }
  void point_id_in_polygon_updated(const std::size_t& p_id, const std::size_t& i1, const std::size_t& i2) final
  {
    std::cout << "In the polygon " << p_id << ", the index " << i1 << " has been replaced by " << i2 << "." << std::endl;
  }
  void polygon_orientation_reversed(const std::size_t& p_id) final
  {
    std::cout << "The polygon " << p_id << " has been reversed." << std::endl;
  }
};
```

File: Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPOrientationVisitor.h [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+/// \cgalHasModel `CGAL::Polygon_mesh_processing::Default_orientation_visitor`.
+
+class PMPOrientationVisitor{
+public:
+
+/// @name Functions used to report non-manifold simplices
+/// @{
+  /// is called each time an edge appears in more than 2 polygons.
+  /// `id1` and `id2` are the ids, from the input point range, of the endpoints of the edge.
+  void non_manifold_edge(const std::size_t & id1, const std::size_t& id2);
```
Comment (MaelRL) [inline Polygon_mesh_processing/doc/Polygon_mesh_processing/Concepts/PMPOrientationVisitor.h:19]: why const& for std::size_t everywhere?

**Comment by MaelRL** [comment]
> I think we should have a function `non_manifold_vertex()` that reports each non-manifold vertex (edge and vertex only as already discussed), and do not only rely on `duplicated_vertex()`. I think all vertex should be replaced by point. @MaelRL, do you have an opinion?

The point vs vertex is indeed awkward at the moment. Maybe it should rather be vertex everywhere? If you have point (geometry) vs edge (combinatorics), it is also inconsistent.

**Comment by sloriot** [comment]
Successfully tested in CGAL-5.4-Ic-38

**Pull request merged**